### PR TITLE
feat(models): remap supports keyword arguments + additional remaps: atanh, asinh, power, affine, displace_boundary_atoms

### PR DESCRIPTION
**Pull request by elkir**
## Description
This allows remap processors to pass parameters from the config to the transform functions/converters. Several remap converters are implemented: renormalised atanh, asinh, pure power transform, affine transform and a special transform that shifts boundary atomic values  away from the fixed bounds for easier modelling

#### `method_kwargs`
Many transforms are a family of functions with input parameters; this change allows us to specify these parameters from the config.

#### Extra transforms
For variables such as tcc, there is a lot of exact atomic mass at 0 and 1, with only a fraction of the values being in between, often clumped to the endpoints. This can be solved by an atanh transform distributing the values equally through the interval, together with the dispalce_boundary_atoms: shifting the boundaries away and then clamping them back to the boundary.

Similarly, for tp, this is usually highly skewed with a lot of mass at exactly 0. A power transform, together with a left boundary atom shift, obtains a region for modelling dry conditions and a reasonably Gaussian wet distribution.

See examples in [Preprocessing — Anemoi Models 0.0.0.post58 documentation](https://anemoi--973.org.readthedocs.build/projects/models/en/973/modules/preprocessing.html)

##  Additional notes ##
- This also cleans up old unused mappings from the deprecated multimapper, which are no longer useful

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

----
📚 Documentation preview 📚: https://anemoi-models--973.org.readthedocs.build/en/973/

----
📚 Documentation preview 📚: https://anemoi-training--973.org.readthedocs.build/en/973/

----
📚 Documentation preview 📚: https://anemoi-graphs--973.org.readthedocs.build/en/973/

### CODE REVIEW

File: models/docs/modules/preprocessing.rst [hunk 1]
```diff
@@ -25,6 +25,120 @@ the following classes:
    :no-undoc-members:
    :show-inheritance:
 
+**********
+Remapper
+**********
+
+The remapper module is used to do online transformations of the data using a set of predefined transforms and their inverses. This process is crucial for variables with pathological distributions, such as variables with a sharp peaks, long tails or other non-Gaussian shapes. It is especially important for diffusion models where the data distribution interacts with the noise distribution.
```
Comment (sahahner) [inline models/docs/modules/preprocessing.rst]:
```suggestion
The remapper module is used to do in-place transformations of the data using a set of predefined transforms and their inverses. This process is crucial for variables with pathological distributions, such as variables with sharp peaks, long tails or other non-Gaussian shapes. It is especially important for diffusion models where the data distribution interacts with the noise distribution.
```

File: models/docs/modules/preprocessing.rst [hunk 2]
```diff
@@ -25,6 +25,120 @@ the following classes:
...
+
+Example configuration:
+
+.. code:: yaml
+
+   data:
+      processors:
+         normalizer:
+           _target_: anemoi.models.preprocessing.multi_dataset_normalizer.TopNormalizer
+           _convert_: all
```
Comment (sahahner) [inline models/docs/modules/preprocessing.rst]: Is the `_convert_: all` option necessary?

File: models/docs/modules/preprocessing.rst [hunk 3]
```diff
@@ -25,6 +25,120 @@ the following classes:
...
+   Atanh remapper transform function examples.
+
+Example configuration:
+
+.. code:: yaml
+
+   data:
+      processors:
+         normalizer:
+           _target_: anemoi.models.preprocessing.multi_dataset_normalizer.TopNormalizer
```
Comment (sahahner) [inline models/docs/modules/preprocessing.rst]: I don't think we have a `multi_dataset_normalizer` in anemoi-models at the moment?

File: models/src/anemoi/models/preprocessing/mappings.py [hunk 4]
```diff
@@ -15,55 +17,224 @@ def noop(x):
...
+            ).all(), (
+                "input x must me greater or equal to 0, or use with clip_negative=True to clip negative values to 0"
+            )
+
+    # Apply transformation
     if lambd == 0:
-        return torch.log(x)
-    return (torch.pow(x, lambd) - 1) / lambd
+        return torch.log_(x)
+    return x.pow_(lambd).sub_(1.0).div_(lambd)
```
Comment (OpheliaMiralles) [inline models/src/anemoi/models/preprocessing/mappings.py:115]: What is the difference between torch log_ and log?

Comment (elkir) [reply models/src/anemoi/models/preprocessing/mappings.py:115]: All the underscore-terminated functions (`log_`,...) operate in-place on the tensor that was given, keeping fixed memory.

File: models/docs/modules/preprocessing.rst (hunk 2 above)
Comment (sahahner) [reply models/docs/modules/preprocessing.rst]: I have unresolved this conversation. Is the `_convert_: all` option necessary? We are not using this option for the other preprocessors, so I wold like to understand why/if it is necessary here.

As far as I understand not using `_convert_: all` is the recommended default.

File: models/docs/modules/preprocessing.rst (hunk 3 above)
Comment (sahahner) [reply models/docs/modules/preprocessing.rst]: There is no multi_dataset_normalizer.TopNormalizer in anemoi

File: models/docs/modules/preprocessing.rst [hunk 5]
```diff
@@ -25,6 +25,120 @@ the following classes:
...
+           _convert_: all
+           config:
+             default: "mean-std"
+             max: ["tp","tcc"]
+         remapper1:
+           _target_: anemoi.models.preprocessing.remapper.Remapper
+           _convert_: all
```
Comment (sahahner) [inline models/docs/modules/preprocessing.rst]: Is the _convert_: all option necessary for the new remapper configuration? 

We should avoid using _convert_: all because it strips away Hydra/OmegaConf features (like interpolation, type safety, and structured configs) and turns everything into plain Python objects. This makes configs less expressive and can hide bugs that Hydra would otherwise catch.

Comment (elkir) [reply models/docs/modules/preprocessing.rst]: Sorry for resolving this comment prematurely. I thought I erased the `__convert__` option when resolving it, but it got lost in the commits somewhere.

The `__convert__` was inherited from a config we were working on, and after checking the documentation, it is indeed not necessary, especially in this form, where we're not storing the normalizer and other configs elsewhere but directly in the list of preprocessors.



Just out of curiosity: was it ever needed for split configs (where we're storing things in separate entries and referring to them through omegaconf), such as the one below, before? Or is this truly coming from some very outdated anemoi code?
```
data:
  normalizer:
    default: "mean-std"
    max: [tp]
  remapper:
    default: "none"
    boxcox-rescaled: [tp]
    method_kwargs: 
      boxcox-rescaled:
        lambd: 0.05
        symmetrical_bound: 1.0
  processors:
    normalizer:
      _target_: anemoi.models.preprocessing.multi_dataset_normalizer.TopNormalizer
      _convert_: all
      config: ${data.normalizer}
    remapper:
      _target_: anemoi.models.preprocessing.ds_remapper.TopRemapper
      _convert_: all
      config: ${data.remapper}
```

File: models/docs/modules/preprocessing.rst (hunk 3 above)
Comment (elkir) [reply models/docs/modules/preprocessing.rst]: I agree, see other comment. Hopefully, the current version has corrections to these now?

File: models/docs/modules/preprocessing.rst (hunk 2 above)
Comment (elkir) [reply models/docs/modules/preprocessing.rst]: roselved in another comment

File: models/docs/modules/preprocessing.rst (hunk 5 above)
Comment (sahahner) [reply models/docs/modules/preprocessing.rst]: Thank you for updating that. I define my processors configurations also in separate entries in the config files and do not use the convert entry.

File: models/src/anemoi/models/preprocessing/mappings.py [hunk 6]
```diff
@@ -15,55 +17,224 @@ def noop(x):
...
+            "Power transform input x must satisfy x >= 0, or use with clip_negative=True to clip negative values to 0."
+        )
+    if tangent_linear_above_one:
+        lin_branch = x.clone().mul_(lambd).add_(1.0 - lambd)
+        pow_branch = x.pow_(lambd)
+        return torch.where(x > 1.0, lin_branch, pow_branch)
+    return x.pow_(lambd)
+
+
+def inverse_power_transform(x, lambd=0.33, tangent_linear_above_one=False):
```
Comment (Lun4m) [inline models/src/anemoi/models/preprocessing/mappings.py:158]: This is a mirror of the comment I left on another PR since they have the same issue.

You are missing the `clip_negative` argument here.

It seems like in the way this is implemented you need the same arguments for both forward and backward transforms, but I think it'd make more sense to have a class for each mapper with forward and backward methods so you can "share" the arguments more explicitly?

This can always be left for a future refactor though.

Comment (sahahner) [reply models/src/anemoi/models/preprocessing/mappings.py:158]: Thank you for raising this. I think we can indeed focus on that in another PR, maybe even the one that you mention :)

Comment (Lun4m) [reply models/src/anemoi/models/preprocessing/mappings.py:158]: Unfortunately that PR is not targeting `main`, so I guess I'll open an issue when this one lands, if my suggestion makes sense

Comment (sahahner) [reply models/src/anemoi/models/preprocessing/mappings.py:158]: okay, feel free to do that. that sounds good :)

Comment (Lun4m) [reply models/src/anemoi/models/preprocessing/mappings.py:158]: The `clip_negative` argument is still missing though, so maybe let's wait for that before resolving? The training would outright crash without it, if you have it set in your config

Comment (Lun4m) [reply models/src/anemoi/models/preprocessing/mappings.py:158]: @sahahner I'm sorry if my first comment was not super clear on the fact there was a bug, but it got merged in :melting_face:

Comment (sahahner) [reply models/src/anemoi/models/preprocessing/mappings.py:158]: Hi, I saw that comment too late. Sorry about that. Could you please open an issue and tag @elkir as well? This should be a small change.

Comment (elkir) [reply models/src/anemoi/models/preprocessing/mappings.py:158]: This has been resolved here https://github.com/ecmwf/anemoi-core/pull/1031

Apparently, some OpenClaw agent found this little bug
